src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -176,22 +180,55 @@ private bool TrySteal(out T result, bool take)
                 CDSCollectionETWBCLProvider.Log.ConcurrentBag_TryPeekSteals();
             }
 
-            // If there's no local queue for this thread, just start from the head queue
-            // and try to steal from each queue until we get a result.
-            WorkStealingQueue localQueue = GetCurrentThreadWorkStealingQueue(forceCreate: false);
-            if (localQueue == null)
+            while (true)
             {
-                return TryStealFromTo(_workStealingQueues, null, out result, take);
-            }
+                // We need to track whether any lists transition from empty to non-empty both before
+                // and after we attempt the steal in case we don't get an item:
+                //
+                // If we don't get an item, we need to handle the possibility of a race condition that led to
+                // an item being added to a list after we already looked at it in a way that breaks
+                // linearizability.  For example, say there are three threads 0, 1, and 2, each with their own
+                // list that's currently empty.  We could then have the following series of operations:
+                // - Thread 2 adds an item, such that there's now 1 item in the bag.
+                // - Thread 1 sees that the count is 1 and does a Take. Its local list is empty, so it tries to
+                //   steal from list 0, but it's empty.  Before it can steal from Thread 2, it's pre-empted.
+                // - Thread 0 adds an item.  The count is now 2.
+                // - Thread 2 takes an item, which comes from its local queue.  The count is now 1.
+                // - Thread 1 continues to try to steal from 2, finds it's empty, and fails its take, even though
+                //   at any given time during its take the count was >= 1.  Oops.
+                // This is particularly problematic for wrapper types that track count using their own synchronization,
+                // e.g. BlockingCollection, and thus expect that a take will always be successful if the number of items
+                // is known to be > 0.
+                //
+                // We work around this by looking at the number of times any list transitions from == 0 to > 0,
+                // checking that before and after the steal attempts.  We don't care about > 0 to > 0 transitions,
+                // because a steal from a list with > 0 elements would have been successful.
+                long initialEmptyToNonEmptyCounts = Interlocked.Read(ref _emptyToNonEmptyListTransitionCount);
```

Comment:
>initialEmptyToNonEmptyCounts [](start = 21, length = 28)

Is it possible in some worst case this operation can starve? I mean while executing here, other threads can push items (which change _emptyToNonEmptyListTransitionCount) and then some other third thread Take the pushed item and this operation continue to happen without having this current thread to get a chance to steal any item and in same time _emptyToNonEmptyListTransitionCount keep changing? I know this is hypothetical but it may be possible.

Reply:
Yes, it's possible, but the chances of that happening are small, and get smaller and smaller the more times it would need to loop.  It's also true with ConcurrentBag in netfx today, as this is just adopting the same scheme it has.